File: src/symbol_table.cpp
#include "symbol_table.h"

SymbolTable::SymbolTable() {

}

void SymbolTable::add(string symbol, int line, int column) {
    if (symbol_table.find(symbol) == symbol_table.end()) {
        OccuranceVector v;
        symbol_table.insert({symbol, v});
    }
    symbol_table[symbol].push_back(LineColumnPair(line, column));
}

OccuranceVector SymbolTable::get(string symbol) {
    return symbol_table[symbol];
}

File: compiler/symbol_table.py
# TODO: Esta tabela de símbolos é muito simples e incompleta.
class SymbolTable():
    _instance = None

    # Singleton
    def __new__(cls):           
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance


    def __init__(self) -> None:
        if not hasattr(self, "_initialized"):
            self.__table: dict[str, dict[str,]] = {}
            self._initialized = True

    def add(self, symbol: str, line: int, column: int):
        entry = self.__table.get(symbol, None)
        if entry:
            self.__table[symbol]["occurrences"].append((line,column))
        else:
            self.__table[symbol] = {
                "occurrences": [(line, column)]
            }
            
    def get(self, symbol: str):
        return self.__table.get(symbol, None)

    def get_all_occurrences(self):
        return self.__table
    
    